File: PS풀이저장소/DP/P12865(평범한 배낭)/P12865.py
from sys import stdin
stdin = open('./input.txt', 'r')

### 1차원 배열
n, k = map(int, stdin.readline().split())
weights = []
values = []

for _ in range(n):
  a, b = map(int, stdin.readline().split())
  weights.append(a)
  values.append(b)

dp = [0] * (k+1)

for i in range(1, n+1):
  w = weights[i-1]
  v = values[i-1]

  for j in range(k, w-1, -1):
    dp[j] = max(dp[j], dp[j-w] + v)
print(dp[k])



### 2차원 배열
n, k = map(int, stdin.readline().split())
weights = []
values = []

for _ in range(n):
  a, b = map(int, stdin.readline().split())
  weights.append(a)
  values.append(b)

dp = [[0] * (k+1) for _ in range(n+1)]

for i in range(1, n+1):
  w = weights[i-1]
  v = values[i-1]

  for j in range(k+1):
    if j < w:
      dp[i][j] = dp[i-1][j]
    else:
      dp[i][j] = max(dp[i-1][j], dp[i-1][j-w] +v)

print(dp[n][k])

File: JS_code/dp/P12865(평범한 배낭)/P12865.js
// P12865. 평범한 배낭
const fs = require('fs');
const input = fs.readFileSync('./input.txt').toString().trim().split('\n');

const [n, k] = input
  .shift()
  .split(' ')
  .map((el) => +el);
const weights = new Array(n).fill(0);
const values = new Array(n).fill(0);
const dp = new Array(k + 1).fill(0);

for (let i = 0; i < n; i++) {
  const [w, v] = input
    .shift()
    .split(' ')
    .map((el) => +el);
  weights[i] = w;
  values[i] = v;
}

for (let i = 0; i < n; i++) {
  const w = weights[i];
  const v = values[i];

  for (let j = k; j >= w; j--) {
    dp[j] = Math.max(dp[j], dp[j - w] + v);
  }
}

console.log(dp[k]);
